interfaces/apparmor/apparmor.go

```diff
@@ -83,22 +85,49 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
 	return nil
 }
 
-// UnloadProfiles removes the named profiles from the running kernel.
-//
-// The operation is done with: apparmor_parser --remove $names...
-// The binary cache file is removed from /var/cache/apparmor
+// UnloadProfiles is meant to remove the named profiles from the running
+// kernel and then remove any cache files. Importantly, we can only unload
+// profiles when we are sure there are no lingering processes from the snap
+// (ie, forcibly stop all running processes from the snap). Otherwise, any
+// running processes will become unconfined. Since we don't have this guarantee
+// yet, leave the profiles loaded in the kernel but remove the cache files from
+// the system so the policy is gone on the next reboot.
 func unloadProfiles(names []string, cacheDir string) error {
 	if len(names) == 0 {
 		return nil
 	}
 
-	args := []string{"--remove"}
-	args = append(args, names...)
-	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
-	if err != nil {
-		return fmt.Errorf("cannot unload apparmor profile: %s\napparmor_parser output:\n%s", err, string(output))
+	/* TODO: uncomment when no lingering snap processes is guaranteed
```

Comment:
This is a very interesting observation. While we knew for a while that apparmor profiles are not really unloaded effectively I was not aware of the semantics of removing a profile while a process is running. Are you certain that the kernel behaves this way? Is the kernel keeping a refcount on each profile and thus allowing us to release the profile but keep it alive internally while processes are running?

This puts an interesting spin on the cgroup notification work.

Reply:
> This is a very interesting observation. While we knew for a while that apparmor profiles are not really unloaded effectively I was not aware of the semantics of removing a profile while a process is running. Are you certain that the kernel behaves this way? Is the kernel keeping a refcount on each profile and thus allowing us to release the profile but keep it alive internally while processes are running?
> 
> This puts an interesting spin on the cgroup notification work.

@zyga - yes which is the behavior for as long as I've used apparmor. You can test this yourself by installing hello-world, running hello-world.sh and trying to access a file (eg, `cat /etc/fstab`), removing the profile and trying to access the file, which you now can:

```
# terminal 1
$ hello-world.sh
...
bash-4.3$ cat /etc/fstab
cat: /etc/fstab: Permission denied
```

```
# terminal 2
$ ps auxxxZ|grep /hello-world
snap.hello-world.sh (enforce)   jamie    21688  0.0  0.0   9608  2620 pts/6    S    08:20   0:00 /bin/bash /snap/hello-world/27/bin/sh
$ sudo apparmor_parser --remove /var/lib/snapd/apparmor/profiles/snap.hello-world.sh
$ ps auxxxZ|grep /hello-world
unconfined                      jamie    21688  0.0  0.0   9608  2620 pts/6    S    08:20   0:00 /bin/bash /snap/hello-world/27/bin/sh
```

```
# terminal 1
bash-4.3$ cat /etc/fstab
# /etc/fstab: static file system information.
...
```

Reply:
I find this terribly un-linux-y but now that we (all) know about it we can work around it. I think this puts the application lifecycle tracking more into "essential" than it was before.